pkg/auth/github.go

```diff
@@ -0,0 +1,172 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// codeResponse represents response body from the Github API
+type codeResponse struct {
+	DeviceCode     string `json:"device_code"`
+	UserCode       string `json:"user_code"`
+	VerficationURI string `json:"verification_uri"`
+	Interval       int    `json:"interval"`
+}
+
+// Uniquely identifies us as a GitHub app.
+// This does not need to be obfuscated because it is publicly available
+// to anyone who does an OAuth request via wego.
+// See the auth ADR for more details:
+// https://github.com/weaveworks/weave-gitops/blob/main/doc/adr/0005-wego-core-auth-strategy.md#design
+const WeGOGithubClientID = "edcb13588d46f254052c"
+
+// Encapsulate shared logic between doCodeRequest and doAuthRequest
+func doRequest(req *http.Request, client *http.Client) ([]byte, error) {
+	req.Header.Set("Accept", "application/json")
+
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		// err is falsey even on 4XX or 5XX
+		return nil, err
+	}
+
+	rb, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return rb, nil
+}
+
+// doCodeRequest does the initial request of the Device Flow
+func doCodeRequest(client *http.Client, scope string) (*codeResponse, error) {
+	query := url.Values.Encode(map[string][]string{
+		"client_id": {WeGOGithubClientID},
+		"scope":     {scope},
+	})
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("https://github.com/login/device/code?%s", query), nil)
```

Comment:
Given that we're using a constant for the `accessTokenUrl`, should we be using one for the device code url?